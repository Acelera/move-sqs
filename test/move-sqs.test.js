const AWS = require('aws-sdk');
const { mockAwsServiceMethod } = require('jest-mock-aws');
const moveSqs = require('../src/move-sqs');
const fixtureFactory = require('./fixture-factory');

describe('move-sqs', () => {
  let input;
  let receiveMessage;
  let sendMessage;
  let deleteMessage;

  beforeEach(() => {
    input = fixtureFactory.createInput();
    receiveMessage = mockAwsServiceMethod(AWS, 'SQS', 'receiveMessage');
    sendMessage = mockAwsServiceMethod(AWS, 'SQS', 'sendMessage');
    deleteMessage = mockAwsServiceMethod(AWS, 'SQS', 'deleteMessage');
    jest.mock('');
  });

  const mockOneMessageOnSourceQueue = () => {
    receiveMessage.mockResolvedValueOnce({
      Messages: [
        {
          Body: JSON.stringify({ hello: 'world' }),
          ReceiptHandle: 'receiptHandle001',
        },
      ],
    });
    receiveMessage.mockResolvedValueOnce({});
  };

  it('should throw an error when required arguments are missing', async () => {
    const expectedError = new Error(
      'Missing required arguments: source-queue-url, destination-queue-url, region, access-key-id, secret-access-key',
    );
    await expect(moveSqs({})).rejects.toEqual(expectedError);
  });

  it('should receive the messages', async () => {
    mockOneMessageOnSourceQueue();
    await moveSqs(input);
    expect(receiveMessage.mock.calls).toMatchSnapshot();
  });

  it('should send a transformed message to the destination queue when input has `mergeJsonMessageWith` specified', async () => {
    mockOneMessageOnSourceQueue();
    input.mergeJsonMessageWith = JSON.stringify({ update: 'my update' });
    await moveSqs(input);
    expect(sendMessage.mock.calls).toMatchSnapshot();
  });

  it('should send the original message to the destination queue when input has no `mergeJsonMessageWith` specified', async () => {
    mockOneMessageOnSourceQueue();
    delete input.mergeJsonMessageWith;
    await moveSqs(input);
    expect(sendMessage.mock.calls).toMatchSnapshot();
  });

  it('should delete the message from the source queue', async () => {
    mockOneMessageOnSourceQueue();
    await moveSqs(input);
    expect(deleteMessage.mock.calls).toMatchSnapshot();
  });
});
