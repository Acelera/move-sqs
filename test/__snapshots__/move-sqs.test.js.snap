// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`move-sqs should delete the message from the source queue 1`] = `
Array [
  Array [
    Object {
      "QueueUrl": "https://sqs.eu-west-1.amazonaws.com/123456789012/the-source-queue",
      "ReceiptHandle": "receiptHandle001",
    },
  ],
]
`;

exports[`move-sqs should receive the messages 1`] = `
Array [
  Array [
    Object {
      "MaxNumberOfMessages": 10,
      "QueueUrl": "https://sqs.eu-west-1.amazonaws.com/123456789012/the-source-queue",
      "VisibilityTimeout": 10,
    },
  ],
  Array [
    Object {
      "MaxNumberOfMessages": 10,
      "QueueUrl": "https://sqs.eu-west-1.amazonaws.com/123456789012/the-source-queue",
      "VisibilityTimeout": 10,
    },
  ],
]
`;

exports[`move-sqs should send a transformed message to the destination queue when input has \`mergeJsonMessageWith\` specified 1`] = `
Array [
  Array [
    Object {
      "MessageBody": "{\\"hello\\":\\"world\\",\\"update\\":\\"my update\\"}",
      "QueueUrl": "https://sqs.eu-west-1.amazonaws.com/123456789012/the-destination-queue",
    },
  ],
]
`;

exports[`move-sqs should send the original message to the destination queue when input has no \`mergeJsonMessageWith\` specified 1`] = `
Array [
  Array [
    Object {
      "MessageBody": "{\\"hello\\":\\"world\\"}",
      "QueueUrl": "https://sqs.eu-west-1.amazonaws.com/123456789012/the-destination-queue",
    },
  ],
]
`;
